### PR TITLE
feat: Workflow that runs each day and creates new issues for failing AVM pipelines

### PR DESCRIPTION
**Pull request by rahalan**
If a AVM pipeline fails, a issue will be created and the failed run is added to the issue.

### CODE REVIEW

File: avm/utilities/pipelines/platform/Set-IssueForWorkflow.ps1
Comment (eriqua) [inline avm/utilities/pipelines/platform/Set-IssueForWorkflow.ps1:1]: Similarly to other platform pipelines, suggest renaming file and function to something like Set-AvmGithubIssueForWorkflow

File: .github/workflows/avm.platform.manage-issue-for-failing-pipelines.yml
Comment (eriqua) [inline .github/workflows/avm.platform.manage-issue-for-failing-pipelines.yml:1]: Suggest to keep file name shorter to better render in GH UI. For example avm.platform.manage-workflow-issue

File: .github/workflows/avm.platform.manage-issue-for-failing-pipelines.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+name: ".Platform: Manage issues for failing pipelines"
```
Comment (eriqua) [inline .github/workflows/avm.platform.manage-issue-for-failing-pipelines.yml:1]: Please name after the file name. E.g. after renaming the yml file, depending on the name chosen, replace with something like
```suggestion
name: "avm.platform.manage-workflow-issue"
```

File: .github/workflows/avm.platform.manage-issue-for-failing-pipelines.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+      - env:
+          GH_TOKEN: ${{ github.token }}
+        name: Manage issues
+        shell: pwsh
+        run: |
+          # Load used functions
+          . (Join-Path $env:GITHUB_WORKSPACE 'avm' 'utilities' 'pipelines'  'platform' 'Set-IssueForWorkflow.ps1')
```
Comment (eriqua) [inline .github/workflows/avm.platform.manage-issue-for-failing-pipelines.yml:23]:
```suggestion
          . (Join-Path $env:GITHUB_WORKSPACE 'avm' 'utilities' 'pipelines' 'platform' 'Set-IssueForWorkflow.ps1')
```

File: avm/utilities/pipelines/platform/Set-IssueForWorkflow.ps1 [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    [int] $LimitNumberOfRuns = 100,
+
+    [Parameter(Mandatory = $false)]
+    [int] $LimitInDays = 2,
+
+    [Parameter(Mandatory = $false)]
+    [String[]] $IgnoreWorkflows = @()
+  )
+
+  $issues = gh issue list --state open --label 'automation,bug' --json 'title,url,body,comments' --repo $Repo | ConvertFrom-Json -Depth 100
```
Comment (eriqua) [inline avm/utilities/pipelines/platform/Set-IssueForWorkflow.ps1:44]: We should check with @matebarabas here which are the labels involved for gh issues managed via automation and related to failing workflows

Comment (rahalan) [reply avm/utilities/pipelines/platform/Set-IssueForWorkflow.ps1]: I changed for now to Type: AVM and Type: Bug

File: .github/workflows/avm.platform.manage-issue-for-failing-pipelines.yml
Comment (rahalan) [reply .github/workflows/avm.platform.manage-issue-for-failing-pipelines.yml:1]: done

File: avm/utilities/pipelines/platform/Set-IssueForWorkflow.ps1
Comment (rahalan) [reply avm/utilities/pipelines/platform/Set-IssueForWorkflow.ps1:1]: done

File: avm/utilities/pipelines/platform/Set-IssueForWorkflow.ps1 (hunk 3 above)
Comment (matebarabas) [reply avm/utilities/pipelines/platform/Set-IssueForWorkflow.ps1]: @rahalan, please see (and only use) the approved labels that are available under the "AVM Standard GitHub Labels" dropdown here: https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr23---category-contributionsupport---github-repo-labels

If any additional labels are needed, let's discuss it internally! ;)

Comment (matebarabas) [reply avm/utilities/pipelines/platform/Set-IssueForWorkflow.ps1]: The ones you picked ('Type: AVM :a: :v: :m:' and 'Type: Bug :bug:') seem to be a good start.

File: .github/workflows/avm.platform.manage-workflow-issue.yml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        shell: pwsh
+        run: |
+          # Load used functions
+          . (Join-Path $env:GITHUB_WORKSPACE 'avm' 'utilities' 'pipelines' 'platform' 'Set-AvmGithubIssueForWorkflow.ps1')
+
+          $functionInput = @{
+            Repo              = "${{ github.repository_owner }}/${{ github.event.repository.name }}"
+            LimitNumberOfRuns = 500
+            LimitInDays       = 2
+            IgnoreWorkflows   = @()
```
Comment (eriqua) [inline .github/workflows/avm.platform.manage-workflow-issue.yml:29]: I'd suggest to filter BRM pipelines running on PR and platform ones here

File: avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1 [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  $issuesClosed = 0
+
+  foreach ($run in $runs) {
+    $runStartTime = [Datetime]::ParseExact($run.startedAt, 'MM/dd/yyyy HH:mm:ss', $null)
+
+    if (($run.headBranch -eq 'main') -and ($runStartTime -gt (Get-Date).AddDays(-$LimitInDays)) -and ($IgnoreWorkflows -notcontains $run.workflowName)) {
+      $runId = $run.url.Split('/') | Select-Object -Last 1
+      $runDetails = gh run view $runId --json 'conclusion,number' --repo $Repo | ConvertFrom-Json -Depth 100
+
+      if ($run.workflowName.StartsWith("avm.")) {
```
Comment (eriqua) [inline avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1:58]: I'd suggest to limit to module workflows. Not sure we want to open issues for platform ones

File: avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1 [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+          }
+        }
+      }
+    }
+  }
+
+  foreach ($workflowRun in $workflowRuns.values) {
+    if ($workflowRun.conclusion -eq 'failure') {
+      $issueName = "[Failed pipeline] $($workflowRun.workflowName)"
+      $failedrun = "failed run: $($workflowRun.url)"
```
Comment (eriqua) [inline avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1:81]:
```suggestion
      $failedrun = "Failed run: $($workflowRun.url)"
```

File: avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1 (hunk 5 above)
Comment (rahalan) [reply avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1:58]: Also platform pipelines should be green. Not sure if we should exclude them.

File: .github/workflows/avm.platform.manage-workflow-issue.yml (hunk 4 above)
Comment (rahalan) [reply .github/workflows/avm.platform.manage-workflow-issue.yml:29]: The filter is further down, filtering for the main branch only

File: avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1 (hunk 5 above)
Comment (eriqua) [reply avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1:58]: We had challenges with that in carml and many stayed in a failing state, but let's see what others think as well. @ChrisSidebotham @AlexanderSehr

Comment (AlexanderSehr) [reply avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1:58]: I'd encourage inluding the platform pipelines. @eriqua could you elaborate on the problems you've experienced? Can't recall right now. 

On that note - at least right now they're all green ;) 
image

File: avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1 [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+.EXAMPLE
+Set-AvmGithubIssueForWorkflow -Repo 'owner/repo01' -LimitNumberOfRuns 100 -LimitInDays 2 -IgnoreWorkflows @('Pipeline 01')
+
+Check the last 100 workflow runs in the repository 'owner/repo01' that happened in the last 2 days. If the workflow name is 'Pipeline 01', then ignore the workflow run.
+
+.NOTES
+The function requires GitHub CLI to be installed.
+#>
+function Set-AvmGithubIssueForWorkflow {
+  [CmdletBinding(SupportsShouldProcess, ConfirmImpact = 'Low')]
```
Comment (AlexanderSehr) [inline avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1:29]: Again, why the `ConfirmImpact`? There must be a reason you added it last time and this time again :D please don't keep me in the dark. I'm not saying you should remove it, but would like to understand the reason :)

File: avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1 [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        if ($PSCmdlet.ShouldProcess("Issue [$issueName]", 'Create')) {
+          gh issue create --title "$issueName" --body "$failedrun" --label 'Type: AVM :a: :v: :m:,Type: Bug :bug:' --repo $Repo
+        }
+
+        $issuesCreated++
+      }
+      else {
+        $issue = ($issues | Where-Object { $_.title -eq $issueName })[0]
+
+        if (!$issue.body.Contains($failedrun)) {
```
Comment (AlexanderSehr) [inline avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1:93]:
```suggestion
        if (-not $issue.body.Contains($failedrun)) {
```

Using `!` in PowerShell is considered a bad practice (even though it works)

File: avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1 [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+          if ($issue.comments.length -eq 0) {
+            if ($PSCmdlet.ShouldProcess("Issue [$issueName]", 'Add comment')) {
+              gh issue comment $issue.url --body $failedrun --repo $Repo
+            }
+
+            $issuesCommented++
+          }
+          else {
+            if (!$issue.comments.body.Contains($failedrun)) {
```
Comment (AlexanderSehr) [inline avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1:102]:
```suggestion
            if (-not $issue.comments.body.Contains($failedrun)) {
```
Using `!` in PowerShell is considered a bad practice (even though it works)

File: avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1 (hunk 7 above)
Comment (rahalan) [reply avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1:29]: The reason is simple: copy & paste :-D . I will remove it

File: avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1 (hunk 5 above)
Comment (eriqua) [reply avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1:58]: I remember we had challenges at least with Linter, opening issues although green state or something on that line. For the ones we have in BRM it should be fine. Question though @rahalan , if an avm issue is created for a failing platform pipeline, what would be the behavior of the assign issue pipeline?

Comment (rahalan) [reply avm/utilities/pipelines/platform/Set-AvmGithubIssueForWorkflow.ps1:58]: @eriqua nothing should happen, as the other pipeline looks for "[AVM Module Issue]" in the title

**Review by eriqua — APPROVED** [review]
:ok:

**Pull request merged**